code/parachain/frame/assets-interface/src/lib.rs

```diff
@@ -0,0 +1,344 @@
+
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Asset Nonce
+	#[pallet::storage]
+	#[allow(clippy::disallowed_types)] // Allow for `ValueQuery` because of nonce
+	pub type AssetNonce<T: Config> =
+		StorageValue<_, u64, ValueQuery, Nonce<ZeroInit, SafeIncrement>>;
+
+	/// Fee to pay for new asset registry
+	#[pallet::storage]
+	pub type AssetCreationFee<T: Config> = StorageValue<_, T::Balance, OptionQuery>;
+
+	/// Asset admin
+	#[pallet::storage]
+	pub type AssetAdmin<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, T::AccountId, OptionQuery>;
+
+	/// Is Asset CosmWasm Asset
+	#[pallet::storage]
+	pub type IsCosmWasmAsset<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, (), OptionQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		NewCreationFeeSet(T::Balance),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AssetNotFound,
+		NotAssetAdmin,
+		BadDecimals,
+		TransferError,
+		CreationFeeIsNotSet,
+		NotCosmWasmAsset,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates an asset for bridge transfer
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::register_asset())]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			location: T::Location,
+			name: Option<BiBoundedAssetName>,
+			symbol: Option<BiBoundedAssetSymbol>,
+			decimals: u8,
+			is_cosmwasm: bool,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let protocol_id = (Pallet::<T>::index() as u32).to_be_bytes();
+
+			let asset_id = <T::AssetsRegistry as GenerateAssetId>::generate_asset_id(
+				protocol_id,
+				AssetNonce::<T>::increment().expect("Does not exceed u64::MAX"),
+			);
+			<Self as PermissionlessAsset>::register_asset(
+				who,
+				asset_id,
+				location,
+				name,
+				symbol,
+				decimals,
+				is_cosmwasm,
+			)?;
+			Ok(())
+		}
+
+		/// Creates an asset for bridge transfer
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as Config>::WeightInfo::set_creation_fee())]
+		pub fn set_creation_fee(origin: OriginFor<T>, fee_amount: T::Balance) -> DispatchResult {
+			T::SetFeeOrigin::ensure_origin(origin)?;
+			AssetCreationFee::<T>::put(fee_amount);
+			Self::deposit_event(Event::NewCreationFeeSet(fee_amount));
+			Ok(())
+		}
+
+		/// Update stored asset information.
+		#[pallet::call_index(2)]
+		#[pallet::weight(<T as Config>::WeightInfo::update_asset())]
+		pub fn update_asset(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			name: UpdateValue<Option<BiBoundedAssetName>>,
+			symbol: UpdateValue<Option<BiBoundedAssetSymbol>>,
+			decimals: UpdateValue<Option<u8>>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let asset_admin =
+				AssetAdmin::<T>::try_get(asset_id).map_err(|_| Error::<T>::AssetNotFound)?;
+			ensure!(asset_admin == who, Error::<T>::NotAssetAdmin);
```

Comment:
i am ok with design when we use separate pallet  just to add admin storage and check, so could we made it into AR pallet? i sure we can. for me it would be nicer and simpler design, more like in parity assets. also it will allow some of multusigns made admins too for us.

Reply:
i think we should keep 2 different pallets because our current AR is storing data for tokens from other pallets like pablo. this allows to have protocol id separate from AR itself, we will also have nonce in AR that only relates to permissionless assets, but there are many types of assets stored. if we create assets via pallet_ibc, we will have a different protocol_id and nonce for it. to me, logically, AR is storage of tokens that doesnt depend how they were created, top level of the tree. all other pallets, pablo, pallet-ibc, permissionless assets with cosmwas, have arrows to it and add info to it. original AR pallet I meant to be  modified via sudo/governance only

Reply:
sure, ibc pallets will send its pid to ar. do not see problem having interface for different consumers. why adding burden of yet another pallet maintainance?

Reply:
plese specify exactly what AR, Assets, AssetsInterface, AssetsRouter doing. I am very sure one is not needed.I think really we need only Assets and AssetsRegistry. Like other chains.

Reply:
sure you can retain all 4, but I still see that things intermixed and confused. starting form security model and ending with dx.

Reply:
we can remove Assets or AssetsRouter, AssetsRouter was supposed to substitute Assets but it isnt relevant anymore